### PR TITLE
Update api-go and sdk-go to allow invalid UTF-8 in protos

### PR DESCRIPTION
**Pull request by tdeebswihart**
## What changed?
I've upgraded the version of our Go API in order to allow invalid UTF-8 data in our proto strings.

Please review https://github.com/temporalio/temporal/pull/5476/commits/5cf979058d0897638f0cf80efea5dbfd737c4a97 only

## Why?
The version of gogo/protobuf we used allowed invalid UTF-8 in strings which is a violation of the proto3 spec. We were unaware of this until our new google/protobuf based structs began failing to deserialize data we'd used previously

## How did you test it?
I added a new end-to-end test to verify that we can create, describe, and delete a namespace that contains invalid UTF-8 in multiple attributes

## Potential risks
This is lower risk than commits prior to this as temporal DBs may already contain invalid UTF-8 data

## Is hotfix candidate?
Yes: this will be going out as part of our 1.23.0 release. In fact, the release is blocked on it!

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -41,8 +41,8 @@ require (
 	go.opentelemetry.io/otel/metric v1.22.0
 	go.opentelemetry.io/otel/sdk v1.21.0
 	go.opentelemetry.io/otel/sdk/metric v1.19.0
-	go.temporal.io/api v1.27.1-0.20240222193246-f8d629e57fa1
-	go.temporal.io/sdk v1.26.0-rc.2
+	go.temporal.io/api v1.29.0
+	go.temporal.io/sdk v1.26.0-rc.3.0.20240301193337-cd94ad4806a7
```
Comment (tdeebswihart) [inline go.mod:45]: This will be updated once the related SDK pr has merged

File: Makefile [hunk 2]
```diff
@@ -40,9 +40,8 @@ PERSISTENCE_DRIVER ?= cassandra
 TEMPORAL_DB ?= temporal
 VISIBILITY_DB ?= temporal_visibility
 
-ifdef TEST_TAG
-override TEST_TAG := -tags $(TEST_TAG)
-endif
+BUILD_TAG := -tags protolegacy
+TEST_TAG := $(BUILD_TAG),$(TEST_TAG)
```
Comment (dnr) [inline Makefile:44]: if we actually get a TEST_TAG then this won't work without override (afaik, I just read the docs).
how about
```suggestion
# Always use "protolegacy" tag to allow disabling utf-8 validation on proto messages
# during proto library transition.
ALL_BUILD_TAGS := protolegacy,$(BUILD_TAG)
ALL_TEST_TAGS := $(ALL_BUILD_TAGS),$(TEST_TAG)
BUILD_TAG_FLAG := -tags $(ALL_BUILD_TAGS)
TEST_TAG_FLAG := -tags $(ALL_TEST_TAGS)

# ... use $(BUILD_TAG_FLAG) and $(TEST_TAG_FLAG) below ...
```

File: tests/namespace_delete.go [hunk 3]
```diff
@@ -106,6 +109,51 @@ func (s *namespaceTestSuite) SetupTest() {
 	s.Assertions = require.New(s.T())
 }
 
+func (s *namespaceTestSuite) Test_NamespaceDelete_InvalidUTF8() {
+	ctx, cancel := rpc.NewContextWithTimeoutAndVersionHeaders(10000 * time.Second)
+	defer cancel()
+
```
Comment (dnr) [inline tests/namespace_delete.go:115]: if you want to be really pedantic you can:
```suggestion
  s.False(utf8.Valid([]byte(invalidUTF8)))
```

File: Makefile (hunk 2 above)
Comment (tdeebswihart) [reply Makefile]: Good catch. I've committed this change

File: Makefile [hunk 4]
```diff
@@ -298,32 +302,32 @@ clean-test-results:
 
 build-tests:
 	@printf $(COLOR) "Build tests..."
-	@go test -exec="true" -count=0 $(TEST_DIRS)
+	@go test $(BUILD_TAG_FLAG) -exec="true" -count=0 $(TEST_DIRS)
```
Comment (dnr) [inline Makefile:305]: probably should be TEST_TAG_FLAG to match the flags used when actually running the tests (one purpose of this target is to be a pre-build to get stuff in cache)

**Review by dnr — APPROVED** [review]
one comment on makefile

File: Makefile (hunk 4 above)
Comment (tdeebswihart) [reply Makefile]: Thanks, I've fixed this

File: go.mod (hunk 1 above)
Comment (tdeebswihart) [reply go.mod]: Done

**Pull request merged**